### DISCUSSION TITLE
I cannot make jar2dll C# .NET Framework 4.7.2/4.8

### DISCUSSION THREAD

**Discussion by maksgranko** [ikvmnet/ikvm#582] [Q&A] [answered]
Hello.
I have been trying for about 2.5 days to convert a JAR to a DLL.
With old version of IKVMC (~8.1.5) it was done in 1 click.
I need to use at least version 8.4.4, however I noticed that the difficulties are the same as with version 8.10, so I decided to use it.

I have re-read a lot of things, however, there is no detailed documentation anywhere for this library.

There were attempts to import all DLLs -recurse, or with -reference from directories:
C:/\Windows\Microsoft.NET\Framework64\v4.0.30319 (I noticed that it picks this up on its own, however, I tried manually prescribing it as well)
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\.NETFramework\v4.7.2
or
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.8
Also, I have tried to use some DLLs separately.
The most information I got was from the command:
ikvmc -target:library -debug:full -nostdlib -out:app.dll -runtime: “D:\Downloads\IKVM-core\IKVM-bin\IKVM.Runtime.dll” -r: “D:\Downloads\IKVM-core\bin\IKVM. Java.dll” -r: ‘C:C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2\mscorlib.dll’ app.jar
Conclusion:
```
...lots of Warning...
warning IKVMC0100: Class “javax.mail.Authenticator” not found

*** COMPILER ERROR ***

IKVM.Tools.Importer, Version=8.10.1.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\
4.0.30319.42000 64-bit
System.NullReferenceException: Object reference does not point to an instance of the object. 
In IKVM.Runtime.AttributeHelper.GetEditorBrowsableNever() in /_/src/IKVM.Runtime/AttributeHelper.cs:line 404   
in IKVM.Runtime.RuntimeByteCodeJavaType.JavaTypeImpl.AnnotationBuilder.Link() in /_/src/IKVM.Runtime/RuntimeByteCodeJavaType.JavaTypeImpl.cs:line 1562   
in IKVM.Runtime.RuntimeByteCodeJavaType.JavaTypeImpl.FinishCore() at /_/src/IKVM.Runtime/RuntimeByteCodeJavaType.JavaTypeImpl.cs:line 1270
   in IKVM.Runtime.RuntimeByteCodeJavaType.JavaTypeImpl.Finish() at /_/src/IKVM.Runtime/RuntimeByteCodeJavaType.JavaTypeImpl.cs:line 1214
   in IKVM.Tools.Importer.RuntimeImportByteCodeJavaType.Finish() in /_/src/IKVM.Tools.Importer/RuntimeImportByteCodeJavaType.cs:line 68   
in IKVM.Runtime.RuntimeByteCodeJavaType.JavaTypeImpl.Finish() in /_/src/IKVM.Runtime/RuntimeByteCodeJavaType.JavaTypeImpl.cs:line 1160   
in IKVM.Tools.Importer.RuntimeImportByteCodeJavaType.Finish() in /_/src/IKVM.Tools.Importer/RuntimeImportByteCodeJavaType.cs:line 68   
in IKVM.Runtime.DynamicClassLoader.FinishAll() in /_/src/IKVM.Runtime/DynamicClassLoader.cs:line 414   
in IKVM.Tools.Tools.Importer.CompilerClassLoader.Compile(RuntimeContext context, StaticCompiler compiler, String runtimeAssembly, List`1 optionsList) at /_/src/IKVM.Tools.Importer/CompilerClassLoader.cs:line 2558   
in IKVM.Tools.Tools.Importer.IkvmImporterInternal.Compile(String[] args) in /_/src/IKVM.Tools.Importer/IkvmImporterInternal.cs:line 196   
In IKVM.Tools.Importer.IkvmImporterInternal.Execute(String[] args) in /_/src/IKVM.Tools.Importer/IkvmImporterInternal.cs:line 113
```

Otherwise, with other commands: “No Core Library”

The application itself, which I want to compile into a DLL, contains everything I need already in its package in a single JAR file.

Then, I decided to try using Visual Studio 2022.
I created a .NET Framework 4.8 Console Mode project.
I tried: Downloading NuGet packages via commands and tutorials, I also tried downloading absolutely ALL packages that are available, I tried downloading packages from the package manager.
In the end I got the error: 
```
Exception raised: “System.TypeInitializationException” in IKVM.Java.dll
System.TypeInitializationException: “The ‘<Module>’ type initializer threw an exception.”
InternalException: Could not locate ikvm home path.
 ```
I also tried to compile in Visual Studio 2022 with x64 and with AnyCPU - the result is the same. 
Based on the names, I tried: Add IKVM(JRE/JDK) folder to the project root.
I also tried adding Tools to the IKVM folder.
Based on IKVM home Path, I tried: Move IKVM to C:\Users\%username%, but this also did not work.
Also, I've tried adding to PATH and to “system variables”:
D:\Downloads\IKVM-core\bin\ (JDK or JRE).
D:\Downloads\IKVM\ (Tools)
IKVM_HOME => D:\Downloads\IKVM-core\bin\
JAVA_HOME => D:\Downloads\IKVM-core\bin\
ikvm => D:\Downloads\IKVM\

I also tried variations with and without bin
PATH:
D:\Downloads\IKVM-core\bin\
D:\Downloads\IKVM-core\
D:\Downloads\IKVM\
%IKVM_HOME%

_Text may be distorted due to translation with DeepL
i tried compile sikulix*
in 8.1.5 he drops an error, which says about IKVM-native, in some issue someone says about 8.4.4_

**Who knows, let me know what I'm doing wrong.**

**Comment by wasabii** [comment]
https://github.com/ikvmnet/ikvm?tab=readme-ov-file#usage

**Reply by maksgranko** [reply]
Ok, I'll try again, next, I'll describe everything I did.
I created a new blank .NET Framework 4.7.2 project for better support
Then, I started the package manager console, typed in:

Install-Package IKVM

Added to the project:
Install-Package IKVM.Image.JDK

Everything installed without errors, successfully.
Since I usually only need it (JDK).

Added x64 configuration
Based on the “IkvmReference” section, I understand I need to modify csproj
I added the necessary lines:
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{0A624FE4-7E70-4EBF-B7FF-08E4DC541E74}</ProjectGuid>
    <OutputType>Exe</OutputType>
    <RootNamespace>ConsoleApp2</RootNamespace>
    <AssemblyName>ConsoleApp2</AssemblyName>
    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <Deterministic>true</Deterministic>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="IKVM.ByteCode, Version=9.0.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.ByteCode.9.0.0\lib\net472\IKVM.ByteCode.dll</HintPath>
    </Reference>
    <Reference Include="IKVM.Java, Version=8.10.1.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.8.10.1\lib\net472\IKVM.Java.dll</HintPath>
    </Reference>
    <Reference Include="IKVM.Runtime, Version=8.10.1.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.8.10.1\lib\net472\IKVM.Runtime.dll</HintPath>
    </Reference>
    <Reference Include="Microsoft.Bcl.AsyncInterfaces, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\Microsoft.Bcl.AsyncInterfaces.8.0.0\lib\net462\Microsoft.Bcl.AsyncInterfaces.dll</HintPath>
    </Reference>
    <Reference Include="System" />
    <Reference Include="System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Buffers.4.5.1\lib\net461\System.Buffers.dll</HintPath>
    </Reference>
    <Reference Include="System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Collections.Immutable.8.0.0\lib\net462\System.Collections.Immutable.dll</HintPath>
    </Reference>
    <Reference Include="System.Core" />
    <Reference Include="System.IO.Pipelines, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.IO.Pipelines.8.0.0\lib\net462\System.IO.Pipelines.dll</HintPath>
    </Reference>
    <Reference Include="System.Memory, Version=4.0.1.2, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Memory.4.5.5\lib\net461\System.Memory.dll</HintPath>
    </Reference>
    <Reference Include="System.Numerics" />
    <Reference Include="System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Numerics.Vectors.4.5.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.6.0.0\lib\net461\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.InteropServices.RuntimeInformation, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.InteropServices.RuntimeInformation.4.3.0\lib\net45\System.Runtime.InteropServices.RuntimeInformation.dll</HintPath>
      <Private>True</Private>
      <Private>True</Private>
    </Reference>
    <Reference Include="System.Text.Encodings.Web, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Text.Encodings.Web.8.0.0\lib\net462\System.Text.Encodings.Web.dll</HintPath>
    </Reference>
    <Reference Include="System.Text.Json, Version=8.0.0.4, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Text.Json.8.0.4\lib\net462\System.Text.Json.dll</HintPath>
    </Reference>
    <Reference Include="System.Threading.Tasks.Extensions, Version=4.2.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.4\lib\net461\System.Threading.Tasks.Extensions.dll</HintPath>
    </Reference>
    <Reference Include="System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.ValueTuple.4.5.0\lib\net47\System.ValueTuple.dll</HintPath>
    </Reference>
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <Compile Include="Program.cs" />
    <Compile Include="Properties\AssemblyInfo.cs" />
  </ItemGroup>
  <ItemGroup>
    <None Include="App.config" />
    <None Include="ikvm.properties" />
    <None Include="packages.config" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
</Project>
```
I won't duplicate it, I'll just write the modified part.
```
...
  <ItemGroup>
    <None Include="App.config" />
    <None Include="ikvm.properties" />
    <None Include="packages.config" />
  </ItemGroup>
  **<ItemGroup>
   <IkvmReference Include="sikulixapi-2.0.5-win.jar">
      <AssemblyVersion>1.0.0.0</AssemblyVersion>
   </IkvmReference>
  </ItemGroup>**
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
...
```

Then, I'm back in Visual Studio, and I see it says, what do you need?
Me: Update everything.
Bottom line, the project works as long as I don't use it 
For an example, I type:
`
static void Main(string[] args)
{
    java.lang.System.gc();
}
`
I get an error:
Exception raised: “System.TypeInitializationException” in IKVM.Java.dll
Unhandled exception of type “System.TypeInitializationException” in Unknown module.
The initializer of type “<Module>” threw an exception.

Also, I removed the project I want to process in the DLL, the error remained.
Ok, as described at the end of the documentation, I will try SDK.Maven.

**Reply by maksgranko** [reply]
Are there any Java dependencies on the PC itself?

**Reply by wasabii** [reply]
That's not an SDK project

**Reply by wasabii** [reply]
https://learn.microsoft.com/en-us/dotnet/core/project-sdk/overview

**Reply by maksgranko** [reply]
Ok, I fixed the problem with the project.
I added IkvmReference/MavenReference (swapped them around)


```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{F763624F-915B-42C1-9A5D-E3130AAFEE23}</ProjectGuid>
    <OutputType>Library</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>Sikulix</RootNamespace>
    <AssemblyName>Sikulix</AssemblyName>
    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <Deterministic>true</Deterministic>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\x64\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <DebugType>full</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <LangVersion>7.3</LangVersion>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>5</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
    <OutputPath>bin\x64\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <Optimize>true</Optimize>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <LangVersion>7.3</LangVersion>
    <ErrorReport>prompt</ErrorReport>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="IKVM.ByteCode, Version=9.0.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.ByteCode.9.0.0\lib\net472\IKVM.ByteCode.dll</HintPath>
    </Reference>
    <Reference Include="IKVM.Java, Version=8.10.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.8.10.0\lib\net472\IKVM.Java.dll</HintPath>
    </Reference>
    <Reference Include="IKVM.Runtime, Version=8.10.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58, processorArchitecture=MSIL">
      <HintPath>..\packages\IKVM.8.10.0\lib\net472\IKVM.Runtime.dll</HintPath>
    </Reference>
    <Reference Include="Microsoft.Bcl.AsyncInterfaces, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\Microsoft.Bcl.AsyncInterfaces.8.0.0\lib\net462\Microsoft.Bcl.AsyncInterfaces.dll</HintPath>
    </Reference>
    <Reference Include="System" />
    <Reference Include="System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Buffers.4.5.1\lib\net461\System.Buffers.dll</HintPath>
    </Reference>
    <Reference Include="System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Collections.Immutable.8.0.0\lib\net462\System.Collections.Immutable.dll</HintPath>
    </Reference>
    <Reference Include="System.Core" />
    <Reference Include="System.IO.Pipelines, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.IO.Pipelines.8.0.0\lib\net462\System.IO.Pipelines.dll</HintPath>
    </Reference>
    <Reference Include="System.Memory, Version=4.0.1.2, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Memory.4.5.5\lib\net461\System.Memory.dll</HintPath>
    </Reference>
    <Reference Include="System.Numerics" />
    <Reference Include="System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Numerics.Vectors.4.5.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.6.0.0\lib\net461\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.InteropServices.RuntimeInformation, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.InteropServices.RuntimeInformation.4.3.0\lib\net45\System.Runtime.InteropServices.RuntimeInformation.dll</HintPath>
      <Private>True</Private>
      <Private>True</Private>
    </Reference>
    <Reference Include="System.Text.Encodings.Web, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Text.Encodings.Web.8.0.0\lib\net462\System.Text.Encodings.Web.dll</HintPath>
    </Reference>
    <Reference Include="System.Text.Json, Version=8.0.0.4, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Text.Json.8.0.4\lib\net462\System.Text.Json.dll</HintPath>
    </Reference>
    <Reference Include="System.Threading.Tasks.Extensions, Version=4.2.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.4\lib\net461\System.Threading.Tasks.Extensions.dll</HintPath>
    </Reference>
    <Reference Include="System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.ValueTuple.4.5.0\lib\net47\System.ValueTuple.dll</HintPath>
    </Reference>
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <Compile Include="Properties\AssemblyInfo.cs" />
  </ItemGroup>
  <ItemGroup>
    <None Include="app.config" />
    <None Include="ikvm.properties" />
    <None Include="packages.config" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
</Project>
```

Changes:

```
...
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <IkvmReference Include="sikulixapi-2.0.5-win.jar">
      <AssemblyVersion>1.0.0.0</AssemblyVersion>
      <AssemblyName>MyAssembly</AssemblyName>
      <AssemblyVersion>3.2.1.0</AssemblyVersion>
      <AssemblyFileVersion>3.0.0.0</AssemblyFileVersion>
      <DisableAutoAssemblyName>true</DisableAutoAssemblyName>
      <DisableAutoAssemblyVersion>true</DisableAutoAssemblyVersion>
      <FallbackAssemblyName>sikulixapi</FallbackAssemblyName>
      <FallbackAssemblyVersion>3.1.0.0</FallbackAssemblyVersion>
      <Compile>sikulixapi-2.0.5-win.jar</Compile>
      <Aliases>sikuli</Aliases>
      <Debug>portable</Debug>
    </IkvmReference>
  </ItemGroup>
  <ItemGroup>
    <Compile Include="Properties\AssemblyInfo.cs" />
...
```

and


```
...
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <IkvmReference Include="sikulixapi-2.0.5-win.jar">
      <AssemblyVersion>1.0.0.0</AssemblyVersion>
    </IkvmReference>
  </ItemGroup>
  <ItemGroup>
    <Compile Include="Properties\AssemblyInfo.cs" />
...
```

but, as a result of building the solution there is no dll or jar, also no errors.
I also tried recopying the jar and compiling with ikvmc in this directory, the error described above is still present.

**Reply by wasabii** [reply, marked as answer]
That's still not a SDK style project. We only support SDK style projects, as they've been mainstream for 6 years now.

**Reply by maksgranko** [reply]
Thanks too much! I made it.

**Reply by raravind88** [reply]
> Thanks too much! I made it.

Am new to ikvm & trying to get 'dll from jar' generated in visual studio 2022 using IkvmReference in csproj file. But dll files aren't getting generated and missing reference error is thrown during build. Could you brief on the steps for dll generation during build phase?

**Reply by Furkanakguun** [reply]
I just want to simply convert .class to .dll with no references and by all the default options. I installed IKVM 8.10.1 from nuget and added my .class inside IkvmReference but still doesn't work.
Guys please explain it.

**Reply by GeorgeS2019** [reply]
@maksgranko 
I do not follow what help you to solve your problem and why the problem happened in the first place

=> Just realized => .NET Framework 4.7.2/4.8 <== this is the problem

**Reply by Furkanakguun** [reply]
> Thanks too much! I made it.

How ? can you share a guideline

**Reply by maksgranko** [reply]
download this file and rename to project.csproj

[project.docx](https://github.com/user-attachments/files/16767848/project.docx)

Move to new folder, open them via Visual Studio 2022

**Reply by maksgranko** [reply]
Wasabi gave a huge clue when I was going through the project types. The project needed is missing(probably I didn't search well) in Visual Studio, so I asked ChatGPT to generate it based on the code at the bottom of the page(I only noticed the project piece after reading the documentation a few times, I don't like microsoft docs).
Anyway.
I sent the file above, you need to edit it a bit, I wrote down where in file)
I recommend that you remove the individual packages at once and follow the Readme, starting from https://github.com/ikvmnet/ikvm?tab=readme-ov-file#installation.

Keep in mind, if you have a ton of dependencies with jar, you need to format them correctly, check out https://github.com/ikvmnet/ikvm?tab=readme-ov-file#syntax.


I think wasabi will mind if I create a separate repository with a link to IKVM, so I sent it as a file.

I hope this helps you.

**Reply by wasabii** [reply]
I mean the MS documentation pretty clearly describes how to make a Framework SDK-style project.

You just create a Core project and change it to Framework.

Or you use the dotnet CLI.

Or you copy/paste the text from the MS site into a file.

**Reply by wasabii** [reply]
https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2022

**Reply by maksgranko** [reply]
Yes, you are right. However, not everyone will understand the basic structure of even this documentation)
Also, I thought for the most part that it was about .NET Core and similar projects, in fact it is not quite so.

**Reply by Furkanakguun** [reply]
I want to work with  Windows Forms (.NET Framework).
